lib/Dialect/LLHD/IR/LLHDOps.cpp

```diff
@@ -987,6 +987,195 @@ static LogicalResult verify(llhd::RegOp op) {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// ForOp
+//===----------------------------------------------------------------------===//
+
+void llhd::ForOp::build(OpBuilder &builder, OperationState &result, Value lb,
+                        Value ub, Value step, ValueRange iterArgs) {
+  result.addOperands({lb, ub, step});
+  result.addOperands(iterArgs);
+  for (Value v : iterArgs)
+    result.addTypes(v.getType());
+  Region *bodyRegion = result.addRegion();
+  bodyRegion->push_back(new Block);
+  Block &bodyBlock = bodyRegion->front();
+  bodyBlock.addArgument(lb.getType());
+  for (Value v : iterArgs)
+    bodyBlock.addArgument(v.getType());
+}
+
+static LogicalResult verifyDAG(DenseMap<Block *, bool> &visited, Block *block) {
+  assert(visited.count(block) &&
+         "Visited map has to be initialized for all blocks in the region!");
+
+  if (visited[block])
+    return failure();
+
+  visited[block] = true;
+  for (Block *succ : block->getSuccessors()) {
+    if (failed(verifyDAG(visited, succ)))
+      return failure();
+  }
+  visited[block] = false;
+  return success();
+}
+
+static LogicalResult verify(llhd::ForOp op) {
+  // Check that the body defines as single block argument for the induction
+  // variable.
+  auto &entryBlock = op.getLoopBody().front();
+  if (!entryBlock.getArgument(0).getType().isSignlessInteger())
+    return op.emitOpError(
+        "expected body first argument to be an signless integer argument for "
+        "the induction variable");
+
+  auto opNumResults = op.getNumResults();
+  // If ForOp defines values, check that the number and types of
+  // the defined values match ForOp initial iter operands and backedge
+  // basic block arguments.
+  if (op.getNumIterOperands() != opNumResults)
+    return op.emitOpError(
+        "mismatch in number of loop-carried values and defined values");
+  if (op.getNumRegionIterArgs() != opNumResults)
+    return op.emitOpError(
+        "mismatch in number of basic block args and defined values");
+  auto iterOperands = op.getIterOperands();
+  auto iterArgs = op.getRegionIterArgs();
+  auto opResults = op.getResults();
+  unsigned i = 0;
+  for (auto e : llvm::zip(iterOperands, iterArgs, opResults)) {
+    if (std::get<0>(e).getType() != std::get<2>(e).getType())
+      return op.emitOpError() << "types mismatch between " << i
+                              << "th iter operand and defined value";
+    if (std::get<1>(e).getType() != std::get<2>(e).getType())
+      return op.emitOpError() << "types mismatch between " << i
+                              << "th iter region arg and defined value";
+
```

Comment:
Is this cut and paste from the std Loop ops, because the message is going to be awkward to read.